File: backend/internal/service/skill_service.go
package service

import (
	"errors"

	"github.com/timebankingskill/backend/internal/models"
	"github.com/timebankingskill/backend/internal/repository"
	"gorm.io/gorm"
)

type SkillService struct {
	skillRepo repository.SkillRepositoryInterface
	userRepo  repository.UserRepositoryInterface
}

func NewSkillService(skillRepo repository.SkillRepositoryInterface, userRepo repository.UserRepositoryInterface) *SkillService {
	return &SkillService{
		skillRepo: skillRepo,
		userRepo:  userRepo,
	}
}

// GetAllSkills retrieves all skills with pagination and filters
func (s *SkillService) GetAllSkills(limit, offset int, category, search string, dayOfWeek *int, minRating *float64, location, sortBy string) ([]models.Skill, int64, error) {
	return s.skillRepo.GetAllWithFilters(limit, offset, category, search, dayOfWeek, minRating, location, sortBy)
}

// GetSkillByID retrieves a skill by ID
func (s *SkillService) GetSkillByID(id uint) (*models.Skill, error) {
	return s.skillRepo.GetByID(id)
}

// GetRecommendedSkills retrieves recommended skills
func (s *SkillService) GetRecommendedSkills(limit int) ([]models.Skill, error) {
	if limit <= 0 {
		limit = 5
	}
	return s.skillRepo.GetRecommendations(limit)
}

// GetSkillTeachers retrieves all teachers for a specific skill
func (s *SkillService) GetSkillTeachers(skillID uint) ([]models.UserSkill, error) {
	// Use type assertion to access the concrete method
	if repo, ok := s.skillRepo.(*repository.SkillRepository); ok {
		return repo.GetTeachersBySkillID(skillID)
	}
	return nil, errors.New("repository does not support this operation")
}

// CreateSkill creates a new skill (admin only)
func (s *SkillService) CreateSkill(skill *models.Skill) error {
	// Validate required fields
	if skill.Name == "" {
		return errors.New("skill name is required")
	}
	if skill.Category == "" {
		return errors.New("skill category is required")
	}

	return s.skillRepo.Create(skill)
}

// UpdateSkill updates an existing skill (admin only)
func (s *SkillService) UpdateSkill(id uint, updates *models.Skill) error {
	// Check if skill exists
	existingSkill, err := s.skillRepo.GetByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("skill not found")
		}
		return err
	}

	// Update fields
	if updates.Name != "" {
		existingSkill.Name = updates.Name
	}
	if updates.Category != "" {
		existingSkill.Category = updates.Category
	}
	if updates.Description != "" {
		existingSkill.Description = updates.Description
	}
	if updates.Icon != "" {
		existingSkill.Icon = updates.Icon
	}

	return s.skillRepo.Update(existingSkill)
}

// DeleteSkill deletes a skill (admin only)
func (s *SkillService) DeleteSkill(id uint) error {
	// Check if skill exists
	_, err := s.skillRepo.GetByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("skill not found")
		}
		return err
	}

	return s.skillRepo.Delete(id)
}

// User Skills Management

// AddUserSkill adds a skill that user can teach to their profile
// This allows users to list skills they can teach to others
//
// Flow:
//   1. Validates user exists
//   2. Validates skill exists in master skills
//   3. Validates required fields (level, hourly rate)
//   4. Checks for duplicate skill (user can't add same skill twice)
//   5. Creates user skill record
//
// Skill Details:
//   - Level: beginner, intermediate, advanced, expert
//   - Hourly Rate: Credits per hour (default 1.0, can be adjusted)
//   - Description: What specifically can be taught
//   - Proof: Optional certificate/portfolio link
//   - Availability: Online/offline preferences
//
// Parameters:
//   - userID: ID of user adding the skill
//   - userSkill: Skill details with level, description, rate, etc
//
// Returns:
//   - error: If validation fails, duplicate skill, or database error
func (s *SkillService) AddUserSkill(userID uint, userSkill *models.UserSkill) error {
	// Validate user exists
	_, err := s.userRepo.GetByID(userID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("user not found")
		}
		return err
	}

	// Validate skill exists
	_, err = s.skillRepo.GetByID(userSkill.SkillID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("skill not found")
		}
		return err
	}

	// Set user ID
	userSkill.UserID = userID

	// Validate required fields
	if userSkill.Level == "" {
		return errors.New("skill level is required")
	}
	if userSkill.HourlyRate < 0 {
		return errors.New("hourly rate must be non-negative")
	}

	// Check if user already has this skill
	existing, _ := s.skillRepo.GetUserSkill(userID, userSkill.SkillID)
	if existing != nil {
		return errors.New("user already has this skill")
	}

	return s.skillRepo.CreateUserSkill(userSkill)
}

// GetUserSkills retrieves all skills that user can teach
func (s *SkillService) GetUserSkills(userID uint) ([]models.UserSkill, error) {
	return s.skillRepo.GetUserSkills(userID)
}

// GetUserSkillByID retrieves a user skill by its ID
func (s *SkillService) GetUserSkillByID(id uint) (*models.UserSkill, error) {
	// Use type assertion to access the concrete method
	if repo, ok := s.skillRepo.(*repository.SkillRepository); ok {
		return repo.GetUserSkillByID(id)
	}
	return nil, errors.New("repository does not support this operation")
}

// UpdateUserSkill updates user's skill details
func (s *SkillService) UpdateUserSkill(userID uint, skillID uint, updates *models.UserSkill) error {
	// Get existing user skill
	existing, err := s.skillRepo.GetUserSkill(userID, skillID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("user skill not found")
		}
		return err
	}

	// Update fields
	if updates.Level != "" {
		existing.Level = updates.Level
	}
	if updates.Description != "" {
		existing.Description = updates.Description
	}
	if updates.YearsOfExperience >= 0 {
		existing.YearsOfExperience = updates.YearsOfExperience
	}
	if updates.ProofURL != "" {
		existing.ProofURL = updates.ProofURL
	}
	if updates.ProofType != "" {
		existing.ProofType = updates.ProofType
	}
	if updates.HourlyRate >= 0 {
		existing.HourlyRate = updates.HourlyRate
	}
	existing.OnlineOnly = updates.OnlineOnly
	existing.OfflineOnly = updates.OfflineOnly
	existing.IsAvailable = updates.IsAvailable

	return s.skillRepo.UpdateUserSkill(existing)
}

// DeleteUserSkill removes a skill from user's teaching skills
func (s *SkillService) DeleteUserSkill(userID uint, skillID uint) error {
	// Check if user skill exists
	_, err := s.skillRepo.GetUserSkill(userID, skillID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("user skill not found")
		}
		return err
	}

	return s.skillRepo.DeleteUserSkill(userID, skillID)
}

// Learning Skills Management

// AddLearningSkill adds a skill to user's learning wishlist
func (s *SkillService) AddLearningSkill(userID uint, learningSkill *models.LearningSkill) error {
	// Validate user exists
	_, err := s.userRepo.GetByID(userID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("user not found")
		}
		return err
	}

	// Validate skill exists
	_, err = s.skillRepo.GetByID(learningSkill.SkillID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("skill not found")
		}
		return err
	}

	// Set user ID
	learningSkill.UserID = userID

	// Check if user already wants to learn this skill
	existing, _ := s.skillRepo.GetLearningSkill(userID, learningSkill.SkillID)
	if existing != nil {
		return errors.New("skill already in learning wishlist")
	}

	return s.skillRepo.CreateLearningSkill(learningSkill)
}

// GetLearningSkills retrieves user's learning wishlist
func (s *SkillService) GetLearningSkills(userID uint) ([]models.LearningSkill, error) {
	return s.skillRepo.GetLearningSkills(userID)
}

// DeleteLearningSkill removes a skill from learning wishlist
func (s *SkillService) DeleteLearningSkill(userID uint, skillID uint) error {
	// Check if learning skill exists
	_, err := s.skillRepo.GetLearningSkill(userID, skillID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("learning skill not found")
		}
		return err
	}

	return s.skillRepo.DeleteLearningSkill(userID, skillID)
}

File: frontend/src/lib/services/skill.service.ts
import { apiClient } from '@/lib/api';
import type { Skill, UserSkill, LearningSkill } from '@/types';

export interface SkillsResponse {
    skills: Skill[];
    total: number;
    limit: number;
    offset: number;
}

export interface AddUserSkillRequest {
    skill_id: number;
    level: string;
    description?: string;
    years_of_experience?: number;
    proof_url?: string;
    proof_type?: string;
    hourly_rate?: number;
    online_only?: boolean;
    offline_only?: boolean;
    is_available?: boolean;
}

export interface UpdateUserSkillRequest {
    level?: string;
    description?: string;
    years_of_experience?: number;
    proof_url?: string;
    proof_type?: string;
    hourly_rate?: number;
    online_only?: boolean;
    offline_only?: boolean;
    is_available?: boolean;
}

export interface AddLearningSkillRequest {
    skill_id: number;
    desired_level?: string;
    notes?: string;
    priority?: number;
}

export const skillService = {
    // Get all skills (master list) with filters
    getAllSkills: async (params?: {
        limit?: number;
        offset?: number;
        category?: string;
        search?: string;
        day?: number;
        rating?: number;
        location?: string;
        sort?: string;
    }): Promise<SkillsResponse> => {
        const queryParams = new URLSearchParams();
        if (params?.limit) queryParams.append('limit', params.limit.toString());
        if (params?.offset) queryParams.append('offset', params.offset.toString());
        if (params?.category) queryParams.append('category', params.category);
        if (params?.search) queryParams.append('search', params.search);
        if (params?.day !== undefined) queryParams.append('day', params.day.toString());
        if (params?.rating) queryParams.append('rating', params.rating.toString());
        if (params?.location) queryParams.append('location', params.location);
        if (params?.sort) queryParams.append('sort', params.sort);

        const query = queryParams.toString();
        return apiClient.get<SkillsResponse>(`/skills${query ? `?${query}` : ''}`);
    },

    // Get skill by ID
    getSkillById: async (id: number): Promise<Skill> => {
        return apiClient.get<Skill>(`/skills/${id}`);
    },

    // Get teachers for a specific skill
    getSkillTeachers: async (skillId: number): Promise<UserSkill[]> => {
        return apiClient.get<UserSkill[]>(`/skills/${skillId}/teachers`);
    },

    // Get recommended skills
    getRecommendedSkills: async (limit?: number): Promise<Skill[]> => {
        return apiClient.get<Skill[]>(`/skills/recommended${limit ? `?limit=${limit}` : ''}`);
    },

    // Get current user's teaching skills
    getUserSkills: async (): Promise<UserSkill[]> => {
        return apiClient.get<UserSkill[]>('/user/skills');
    },

    // Add a teaching skill
    addUserSkill: async (data: AddUserSkillRequest): Promise<UserSkill> => {
        return apiClient.post<UserSkill>('/user/skills', data);
    },

    // Update a teaching skill
    updateUserSkill: async (skillId: number, data: UpdateUserSkillRequest): Promise<UserSkill> => {
        return apiClient.put<UserSkill>(`/user/skills/${skillId}`, data);
    },

    // Delete a teaching skill
    deleteUserSkill: async (skillId: number): Promise<void> => {
        return apiClient.delete(`/user/skills/${skillId}`);
    },

    // Get current user's learning skills (wishlist)
    getLearningSkills: async (): Promise<LearningSkill[]> => {
        return apiClient.get<LearningSkill[]>('/user/learning-skills');
    },

    // Add a learning skill
    addLearningSkill: async (data: AddLearningSkillRequest): Promise<LearningSkill> => {
        return apiClient.post<LearningSkill>('/user/learning-skills', data);
    },

    // Delete a learning skill
    deleteLearningSkill: async (skillId: number): Promise<void> => {
        return apiClient.delete(`/user/learning-skills/${skillId}`);
    },
};
